docs: Documentation for Query Acceleration with Blooms (#12486)
grafana/loki: docs/sources/operations/query-acceleration-blooms.md

---
title: Query Acceleration with Blooms (Experimental) 

> menuTitle: Query Acceleration with Blooms (Experimental) 

Reviewer: This is a bit too long for a menuTitle. Since the topic is marked as experimental, I don't think we need to mention it in the navigation.
```suggestion
menuTitle: Query Acceleration with Blooms 
```

and served by the new [Bloom Gateway](#bloom-gateway) component.

## Enable Query Acceleration with Blooms
To start building and using blooms you need to:
- Deploy the [Bloom Compactor](#bloom-compactor) component (as a [microservice][microservices] or via the [SSD][ssd] Backend target) and enable the component in the [Bloom Compactor config][compactor-cfg].
- Deploy the [Bloom Gateway](#bloom-gateway) component (as a [microservice][microservices] or via the [SSD][ssd] Backend target) and enable the component in the [Bloom Gateway config][gateway-cfg].
- Enable blooms filtering and compaction for each tenant individually, or for all of them by default.

```yaml
bloom_compactor:
  enabled: true

bloom_gateway:
  enabled: true
  client:
    addresses: dnssrvnoa+_bloom-gateway-grpc._tcp.bloom-gateway-headless.<namespace>.svc.cluster.local

# Enable blooms filtering and compaction for all tenants by default
limits_config:
  bloom_gateway_enable_filtering: true
  bloom_compactor_enable_compaction: true
```

For more configuration options refer to the [Bloom Gateways][gateway-cfg], [Bloom Compactor][compactor-cfg] and 
[per tenant-limits][tenant-limits] configuration docs. 

> We strongly recommend reading the whole documentation hereafter for this experimental feature before using it.

Reviewer: ```suggestion
We strongly recommend reading the whole documentation for this experimental feature before using it.
```

## Bloom Compactor
The Bloom Compactor component builds blooms from the chunks in the object store. 

> The resulting blooms are grouped in bloom blocks spanning multiple streams (aka series) and chunks from a given day. 

Reviewer: ```suggestion
The resulting blooms are grouped in bloom blocks spanning multiple streams (also known as series) and chunks from a given day. 
```

> To learn more about how blocks and metadata files are organized, check the 

Reviewer: ```suggestion
To learn more about how blocks and metadata files are organized, refer to the 
```

[Building and querying blooms](#building-and-querying-blooms) section below.

Bloom Compactors are horizontally scalable and use a [ring] for sharding tenants and stream fingerprints, 
as well as determining which compactor should apply [blooms retention](#retention). 
Each compactor owns a configurable number of contiguous streams fingerprint ranges for a tenant. 
The compactor builds blooms for all the chunks from the tenant streams whose fingerprint 
falls within its owned key-space ranges.

You can find all the configuration options for this component in the [Configure section for the Bloom Compactor][compactor-cfg].
Refer to the [Enable Query Acceleration with Blooms](#enable-query-acceleration-with-blooms) section below for 

> a configuration snipped enabling this feature.

Reviewer: ```suggestion
a configuration snippet enabling this feature.
```

### Retention

> One Bloom Compactor from all the running will apply retention. Retention is disabled by default.

Reviewer: ```suggestion
One Bloom Compactor from all those running will apply retention. Retention is disabled by default.
```

locally attached SSD disks (NVMe) to increase i/o throughput. 
Multiple directories on different disk mounts can be specified using the `-bloom.shipper.working-directory` [setting][gateway-cfg] 
when using a comma separated list of mount points, for example:
```
-bloom.shipper.working-directory="/mnt/data0,/mnt/data1,/mnt/data2,/mnt/data3"
```

Bloom Gateways need to deal with relatively large files: the bloom filter blocks. 
Even though the binary format of the bloom blocks allows for reading them into memory in smaller pages, 
the memory consumption depends on the amount of pages that are concurrently loaded into memory for processing. 
The product of three settings control the maximum amount of bloom data in memory at any given 
time: `-bloom-gateway.worker-concurrency`, `-bloom-gateway.block-query-concurrency`, and `-bloom.max-query-page-size`.

Example, assuming 4 CPU cores:
```
-bloom-gateway.worker-concurrency=4      // 1x NUM_CORES
-bloom-gateway.block-query-concurrency=8 // 2x NUM_CORES
-bloom.max-query-page-size=64MiB

4 x 8 x 64MiB = 2048MiB
```

Here, the memory requirement for block processing is 2GiB.

> To get the minimum requirements for the Bloom Gateways, you need to double the value

Reviewer: ```suggestion
To get the minimum requirements for the Bloom Gateways, you need to double the value.
```

TSDB index files they were built from. They also contain tombstones for old blocks which are outdated and 
can be deleted in future iterations. Gateways and compactors use these metadata files to discover existing blocks.

Every `-bloom-compactor.compaction-interval`, compactors will load the latest TSDB files for all tenants for
which bloom compaction is enabled, and compare the TSDB files with the latest bloom metadata files. 
If there are new TSDB files or any of them have changed, the compactor will process all the streams and chunks pointed 
by the TSDB file. In case of changes for a previously processed TSDB file, 
compactors will try to reuse blooms from existing blocks instead of building new ones from scratch.

For a given stream, the compactor owning that stream will iterate through all the log lines inside its new 
chunks and build a bloom for the stream. For each log line, we compute its [n-grams](https://en.wikipedia.org/wiki/N-gram#:~:text=An%20n%2Dgram%20is%20a,pairs%20extracted%20from%20a%20genome.) 
and append to the bloom both the hash for each n-gram and the hash for each n-gram plus the chunk identifier. 

> The former allows gateways to skip whole stream while the latter is for skipping individual chunks.

Reviewer: ```suggestion
The former allows gateways to skip whole streams while the latter is for skipping individual chunks.
```

but the longer filtering expressions need to be able to check them against blooms. 
For the example above, where the n-gram length is 3, we need filtering expressions that have at least 3 characters.

### Queries for which blooms are used
Loki will check blooms for any log filtering expression within a query that satisfies the following criteria:
- The filtering expression contains at least as many characters as the n-gram length used to build the blooms.
  - For example, if the n-grams length is 5, the filter `|= "foo"` will not take advantage of blooms but `|= "foobar"` would.
- If the filter is a regex, we use blooms only if we can simplify the regex to a set of simple matchers.
  - For example, `|~ "(error|warn)"` would be simplified into `|= "error" or "warn"` thus would make use of blooms, 
    whereas `|~ "f.*oo"` would not be simplifiable.
- The filtering expression is a match (`|=`) or regex match (`|~`) filter. We don’t use blooms for not equal (`!=`) or not regex (`!~`) expressions.
  - For example, `|= "level=error"` would use blooms but `!= "level=error"` would not.

> - The filtering expression is placed before a [line format expression]({{< relref "../query/log_queries#line-format-expression" >}}).

Reviewer: ```suggestion
- The filtering expression is placed before a [line format expression](https://grafana.com/docs/loki/latest/query/log_queries/#line-format-expression).
```

    the first filter (`|= "level=error"`) will benefit from blooms but the second one (`|= "traceID=3ksn8d4jj3"`) will not.

## Query sharding
Query acceleration does not just happen while processing chunks, 
but also happens from the query planning phase where the query frontend applies [query sharding](https://lokidex.com/posts/tsdb/#sharding). 
Loki 3.0 introduces a new {per-tenant configuration][tenant-limits] flag `tsdb_sharding_strategy` which defaults to computing 
shards as in previous versions of Loki by using the index stats to come up with the closest power of two that would 
optimistically divide the data to process in shards of roughly the same size. Unfortunately, 
the amount of data each stream has is often unbalanced with the rest, 
therefore, some shards end up processing more data than others.

Query acceleration introduces a new sharding strategy: `bounded`, which uses blooms to reduce the chunks to be 
processed right away during the planning phase in the query frontend, 
as well as evenly distributes the amount of chunks each sharded query will need to process.

> [ring]: {{< relref "../get-started/hash-rings" >}}

Reviewer: ```suggestion
[ring]: https://grafana.com/docs/loki/latest/get-started/hash-rings/
```

> [tenant-limits]: {{< relref "../configure/#limits_config" >}}

Reviewer: ```suggestion
[tenant-limits]: https://grafana.com/docs/loki/latest/configure/#limits_config
```

> [gateway-cfg]: {{< relref "../configure/#bloom_gateway" >}}

Reviewer: ```suggestion
[gateway-cfg]: https://grafana.com/docs/loki/latest/configure/#bloom_gateway
```

> [compactor-cfg]: {{< relref "../configure/#bloom_compactor" >}}

Reviewer: ```suggestion
[compactor-cfg]: https://grafana.com/docs/loki/latest/configure/#bloom_compactor
```

> [microservices]: {{< relref "../get-started/deployment-modes/#microservices-mode" >}}

Reviewer: ```suggestion
[microservices]: https://grafana.com/docs/loki/latest/get-started/deployment-modes/#microservices-mode
```

> [ssd]: {{< relref "../get-started/deployment-modes/#simple-scalable" >}}

Reviewer: ```suggestion
[ssd]: https://grafana.com/docs/loki/latest/get-started/deployment-modes/#simple-scalable
```
